crates/core/src/ring/mod.rs

```diff
@@ -271,14 +305,28 @@ impl Ring {
 
         // Get all connected peers through the connection manager
         let connections = self.connection_manager.get_connections_by_location();
-        let peers = connections.values().filter_map(|conns| {
-            use rand::seq::SliceRandom;
-            let conn = conns.choose(&mut rand::thread_rng())?;
-            (!skip_list.has_element(conn.location.peer.clone())).then_some(&conn.location)
-        });
+        let mut candidates: Vec<PeerKeyLocation> = connections
+            .values()
+            .filter_map(|conns| {
+                use rand::seq::SliceRandom;
+                let conn = conns.choose(&mut rand::thread_rng())?;
+                (!skip_list.has_element(conn.location.peer.clone()))
+                    .then_some(conn.location.clone())
+            })
+            .collect();
+
+        // Also consider self if we have a location and aren't in skip list
+        if let Some(own_peer) = self.connection_manager.get_peer_key() {
+            if !skip_list.has_element(own_peer) {
+                let own_pkloc = self.connection_manager.own_location();
+                if own_pkloc.location.is_some() {
+                    candidates.push(own_pkloc);
+                }
+            }
+        }
 
         router
-            .select_k_best_peers(peers, target_location, k)
+            .select_k_best_peers(candidates.iter(), target_location, k)
```

Comment:
we can probably chain the `peers` iterator (instead of collection) with an other iterator which is basically `[own_pkloc]`(avoids collecting and allocating unnecessarily)